Visual Sudoku example

CryoCardiogram opened the pull request:
Here is the visual sudoku advanced example, both in a .py script and a notebook. 
It's not ready/clean yet, there is a lot of room for improvement.

tias commented on examples/advanced/visual_sudoku.py:
```diff
@@ -0,0 +1,281 @@
+    # our NN returns 81 probabilistic vector: an 81x10 matrix
+    return pred.reshape(9,9,10).detach() # reshape as 9x9x10 tensor for easier visualisation
+
+if __name__ == '__main__':
```
create a function `def visual_sudoku()` that you call here (rest is blank).

Put that `def visual_sudoku()` at the top: make your script readable from top to bottom

tias commented on examples/advanced/visual_sudoku.py:
```diff
@@ -0,0 +1,281 @@
+        [e, e, e,  6, e, 8,  e, e, e]
+    ])
+
+    # Emilio: Missing a bit of context here ?
```
idd, also watch your variable names, you call it 'sample_sudoku_grid' which you then feed into a 'sample_...' function which means somethign else.

Also higher up, avoid 'model' for the NN as 'model' in CP/CPMpy is smth different...

tias commented on examples/advanced/visual_sudoku.py:
```diff
@@ -0,0 +1,281 @@
+    vs = sample_visual_sudoku(sample_sudoku_grid)
+
+    # TODO: (Minor) Does not work from outside of folder + what is clf ? (classifier?)
+    model = load_clf(LeNet,'lenet_mnist_e15.pt')
```
do not use 'model', but something descriptive like 'nn'

tias replied:
or clf, given the load_clf function name...

tias commented on examples/advanced/visual_sudoku.py:
```diff
@@ -0,0 +1,281 @@
+    model.solve()
+    solution = dvar.value()
+
+    # TODO: pretty printing ?
```
work on the flow...

You first want to compute the argmax solution, so have only the code for that, and print that result.
Some visu/print/annotation of how it makes a mistake would be good.... (a notebook could be better then? also forces you to be more linear...)

Then, write that you will now make a max likelihood formulation.

Ideally, the reader can see the difference very close to each other, e.g.

m = base_sudoku(grid)
m += (grid[given] == argmax[given])
m.solve()

m2 = base_sudoku(grid)
m2.optimize(np.sum((grid==i)*logprobs[i] for i in range()) # or some fancy one-liner to write this ; )
m2.solve()


THAT would be a great example...

then, advanced edition, can build on m2 perhaps?
while not_unique(grid.value()):
  m2 += # blocking clause
  m2.solve()


I'm not sure if it is doable, but if your 'def visu_sudoku():` would contain this, it would be a very pedagogical example...

CryoCardiogram commented:
I have updated a notebook and deleted the .py script. 

I have reworked the flow, added visualization and verbosity / markdown cells. 
Ready for a final review and/or merge.
